### ISSUE TITLE
Enable SourceLink for the NuGet APIs packages

### ISSUE THREAD

**Issue by lbargaoanu** [NuGet/Home#10463]
If there is already an issue for this, I didn't find it :)

**Comment by zivkan** [comment]
Blocked on https://github.com/dotnet/sourcelink/issues/696

**Comment by zivkan** [comment]
Update for anyone curious. We learned that our pdbs are being generated correctly, but the issue is that Microsoft's symbol server is intentionally stripping out sourcelink information. I assume it's because most Microsoft products are closed source.

What the dotnet org does is:

* Generate snupkgs with portable pdbs
* Publish nupkg+snupkg to nuget.org
* Use pdb2pdb.exe to convert to pdbs to windows ("full") pdbs
* use the converted pdbs to publish to Microsoft's symbol server, which is a requirement for Visual Studio.

This allows customers to use nuget.org's symbol server gets symbols and debug our packages on nuget.org.  The binaries in VS is from a different NuGet build, hence has different pdbs. Therefore, there shouldn't be an issue with the same pdb existing in two places, one with sourcelink, one without.

**Comment by lbargaoanu** [comment]
While interesting, this doesn't help :) I'm using NuGet.Commands 5.8.0. I cannot debug NuGet. The only symbols found are on Microsoft's symbol server, and those don't do the trick. If I disable Microsoft's symbol server (and keep the NuGet symbol server), no symbols are found. So if SourceLink is supposed to work, how's that exactly? :) I think I gave it a fair shot.

**Comment by heng-liu** [comment]
As @zivkan mentioned above, Microsoft's symbol server is intentionally stripping out sourcelink information. That's why you could not debug into it.
After we finish the above steps and publish new versions of NuGet packages, you should be able to debug into that.

**Comment by lbargaoanu** [comment]
Thanks. It wasn't obvious to me what's the current situation (other than what I saw).

**Comment by heng-liu** [comment]
Currently got blocked. 
As I'm not able to generate the right snupkg and nupkg packages for NuGet.Build.Tasks.Pack project.
@nkolev92 @zkat could you help advise how to generate snupkg and nupkg correctly for NuGet.Build.Tasks.Pack?
For now in dev, the symbols.nupkg contains no pdb file(seems not correct either).
After I change the format to snupkg (branch dev-hengliu-CI-enableSourceLink), the pack failed for:
```
"C:\repos\NuGet.Client\src\NuGet.Core\NuGet.Build.Tasks.Pack\NuGet.Build.Tasks.Pack.csproj" (pack target) (1) ->

(GenerateNuspec target) ->

  C:\Program Files\dotnet\sdk\5.0.202\Sdks\NuGet.Build.Tasks.Pack\buildCrossTargeting\NuGet.Build.Tasks.Pack.targets(221,5): error NU5017: Cannot create a package that has no dependencies nor content. [C:\repos\N

uGet.Client\src\NuGet.Core\NuGet.Build.Tasks.Pack\NuGet.Build.Tasks.Pack.csproj]
```

**Comment by nkolev92** [comment]
The trick why things don't work is because we custom pack using: https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj#L175-L190

We also set `IncludeBuildOutput` to `false`, https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.csproj#L14, which basically means that none of the standard collection tasks don't work. 

The target that collects the pdbs is not running because of that, https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L373-L382. 

I tried to play with it by adding the following: 

```xml
  <Target Name="AddSymbolsPerFramework"
        Returns="@(_TargetPathsToSymbols)">
  <PropertyGroup>
    <ILMergeSubpath Condition="'$(IsBuildOnlyXPLATProjects)' != 'true'">ilmerge\</ILMergeSubpath>
  </PropertyGroup>
  <ItemGroup>
    <_TargetPathsToSymbols Include="$(OutputPath)$(ILMergeSubpath)NuGet.Build.Tasks.Pack.pdb">
      <TargetFramework>$(TargetFramework)</TargetFramework>
    </_TargetPathsToSymbols>
  </ItemGroup>
  </Target>

  <Target Name="AddSymbols"
        AfterTargets="_WalkEachTargetPerFramework"
        Returns="@(_TargetPathsToSymbols)">
    <MSBuild
      Projects="$(MSBuildProjectFullPath)"
      Targets="AddSymbolsPerFramework"
      Properties="TargetFramework=%(_TargetFrameworks.Identity);">

      <Output
          TaskParameter="TargetOutputs"
          ItemName="_TargetPathsToSymbols" />
    </MSBuild>
  </Target>
``` 

in the pack project csproj.

That leads to populating `TargetPathsToSymbols` which is what is passed to the PackTask, https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Build.Tasks.Pack/PackTask.cs#L39, but the error still happened. 

Digging further, I see that even if that item is set, the include is conditioned on https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs#L119-L124. 

This doesn't it's impossible customize it, but that our current approach is lacking, https://github.com/NuGet/NuGet.Client/blob/d2bb0de35242b04802c1a7856ebb7718f674565c/src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs#L119-L124. 

I tried setting `IncludeBuildOutput` to `true` and then I arrived to 

```xml
  <Target Name="_GetBuildOutputFilesWithTfm"
          Returns="@(BuildOutputInPackage)">
  <!-- Do nothing!-->
  </Target>

  <Target Name="_GetDebugSymbolsWithTfm"
        Condition="'$(IncludeBuildOutput)' == 'true'"
        Returns="@(_TargetPathsToSymbolsWithTfm)">
    <PropertyGroup>
      <ILMergeSubpath Condition="'$(IsBuildOnlyXPLATProjects)' != 'true'">ilmerge\</ILMergeSubpath>
    </PropertyGroup>
    <ItemGroup>
      <_TargetPathsToSymbolsWithTfm Include="$(OutputPath)$(ILMergeSubpath)NuGet.Build.Tasks.Pack.pdb">
        <TargetFramework>$(TargetFramework)</TargetFramework>
      </_TargetPathsToSymbolsWithTfm>
    </ItemGroup>
  </Target>
```

This does work, but the pdbs are in the wrong folder (lib, not the desktop) folder. 

I am not sure if nuget.org publishing will accept that. 

tldr; 

This is pretty hard. Should we skip it for the pack package?

**Comment by heng-liu** [comment]
Thanks for both @zivkan and @nkolev92's help offline.
We agreed to skip the packing of NuGet.Build.Tasks.Pack.snupkg for now.
When https://github.com/NuGet/Home/issues/10820 is fixed, we will start packing symbol package for project NuGet.Build.Tasks.Pack.

**Comment by zivkan** [comment]
I just had a thought that we might not be able to do this. Our insertions into VS & .NET SDK need to have symbols, so we can't stop publishing symbols for NuGet.Build.Tasks.Pack.dll. As long as we get NuGet.Build.Tasks.Pack.pdb from our build, then it's fine if we don't create a snupkg for this one package, but we need to ensure that we keep the correct pdb for this assembly when we publish internally to symweb.

**Comment by zivkan** [comment]
This PR might unblock our NuGet.Build.Tasks.Pack symbols problem: https://github.com/NuGet/NuGet.Client/pull/4053

**Comment by dominoFire** [comment]
Having issues with ilmerge and portable PDB's. See this issue for more info: https://github.com/dotnet/ILMerge/issues/11

**Comment by dominoFire** [comment]
Neither ilrepack.exe or ilmerge.exe supports Portable PDB's merging. See:

- https://github.com/dotnet/ILMerge/issues/11
- https://github.com/gluck/il-repack/issues/197

Alternative:

- Build NuGet.Client with full PDBs `<DebugType>full<DebugType>` 
- Run pdb2pdb.exe to convert .NET Core full PDBs to Portable PDBs
- Pack nupkg's and snupkg's
- Publish nupkgs and snupkgs to nuget.org
- Publish full PDBs to VS

**Comment by dominoFire** [comment]
Lowering to P3 since you can debug if you debug NuGet from source code and required PDB-converter tools don't work for our use case

**Comment by zivkan** [comment]
Symbols for NuGet SDK packages have been published for version 6.0 and higher. Some earlier versions also have symbols published, but it's proved challenging to easily find all the exact builds for versions without published pdbs, so we couldn't backfill earlier versions. However, NuGet 5.x corresponds to .NET 5.0 which is already out of support at the time I'm writing this.

Symbols have been published to Microsoft's symbol server, not nuget.org's symbol server (blame ilmerge and nuget.exe for making it infeasible to use portable pdbs in our builds, necessary to publish symbols to nuget.org).

Some time during the last 12 months nuget's pdbs stopped stripping SourceLink information. So, NuGet 6.0.0 packages like NuGet.Versioning will have pdbs available, but not source link. I confirmed that 6.2.0 definitely has SourceLink.  Note that due to mistakes made, 6.2.1 has source link, but pointing to a commit hash that we couldn't make public. Hence why I verified that 6.2.0 works as expected.
